### DISCUSSION TITLE
css and other asset resources loading failed

### DISCUSSION THREAD

**Discussion by Harrypotterrrr** [mmistakes/minimal-mistakes#2631] [Q&A]
## Environment

- Minimal Mistakes version: I use [academicpages](https://github.com/academicpages/academicpages.github.io)
- Ruby gem or remote theme version: 
  ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
  gem 2.7.6

- Jekyll version: 3.8.7

- Git repository URL: https://github.com/Harrypotterrrr/Homepage
- GitHub Pages hosted (if yes provide URL to site):
- Operating system: Linux ubuntu6-1 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:39:52 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

## Expected behavior

Hi, I am using [academicpages](https://github.com/academicpages/academicpages.github.io), a repository forked from this minimal-misktake which is very suitable to the academic propose. I am trying to debug and modify my homepage by using `bundle exec jekyll serve --host 0.0.0.0 --port 8445 --incremental` on the remote server. Note that this remote server is the one with inner IP and cannot be accessed from outside, so I use port forward technique to propagate the port to the public server IP by using `ssh -NfR 8445:localhost:8445 root@106.13.xx.xx`. 

I manage to see all the text of my homepage, while I fail to load any CSS and asset images etc. As the below figure shows:
![image](https://user-images.githubusercontent.com/28552583/88963968-3c538a00-d2db-11ea-9d95-589a148561c9.png)

You can see the request is `http://0.0.0.0:8445/assets/css/main.css`. I change `0.0.0.0` to my public server IP `106.13.xx.xx`, that is
 `http://106.13.xx.xx:8445/assets/css/main.css`, I successfully get the `main.css` file as the following image shows, which means port forward success.
![image](https://user-images.githubusercontent.com/28552583/88964189-95bbb900-d2db-11ea-9f79-b84318bebbfc.png)

So I turn to the `_config.yml` to see if I could modify `host` or something else to direct assets request to `106.13.xx.xx` instead of `0.0.0.0` but finally I failed.

Struggling for hours, I have to ask you for help for I am a newbie in front-end tech. 
Hope I make my problem clear and easy to understand. I just wanna test my modification and debug on that remote server, which has an inner IP address and can not be accessed outside. I have no problem with deploying it to Github.
Thanks for your help in advance!

**Comment by iBug** [comment]
This is an upstream issue. Jekyll overrides `site.url` when running in `serve` mode, which by default is `http://127.0.0.1:4000` and changes with your host/port settings.

If you're using SSH forwarding, just launch Jekyll with default host / port and let SSH forward `4000:127.0.0.1:4000` and it should behave normally.

**In short**, drop the `--host` option, it's unnecessary.

**Comment by Harrypotterrrr** [comment]
@iBug Hi, appreciate your instant reply. I tried to use default host and port: `bundle exec jekyll serve --incremental` but nothing seems change. I still fail to get CSS and other assets files.
![image](https://user-images.githubusercontent.com/28552583/89059810-70d74c80-d394-11ea-84a0-8101c3b7e299.png)
I use this command to forward SSH port: `ssh -NfR 4000:localhost:4000 root@106.13.xx.xx `
Here is what I see in my chrome browser. sad.. 😢 
![image](https://user-images.githubusercontent.com/28552583/89060044-db888800-d394-11ea-9219-921a883c6065.png)

**Comment by iBug** [comment]
Are you sure your SSH client is running properly? With a Linux machine, you can check if there's an `ssh` process listening on port 4000 with the command `netstat -tlpn` and look into its output.

In addition, with SSH forwarding enabled, you shouldn't visit the server's address directly, but only <http://127.0.0.1:4000/>, as SSH will handle all the traffic.

Anyways, this is not a theme-specific issue, so I'd suggest that you seek help over on [Jekyll Talk](https://talk.jekyllrb.com/) where people with more expertise can assist you.

**Comment by mmistakes** [comment]
Closing as this is an environment/upstream issue and not something impacted by the theme.

**Comment by Harrypotterrrr** [comment]
@iBug I confirm SSH client runs properly. I use code-server on this inner remote server, and I modify codes remotely on my local PC. I use `netstat- nultp` commands and the process seems to work well as below.
![image](https://user-images.githubusercontent.com/28552583/89061220-ec39fd80-d396-11ea-8bda-2679a3dd3735.png)
I use remote port forwarding, which means I could visit the remote server which has the public IP and request for resources on the remote private server.
This is not a theme-specific issue, anyway, thanks for your reply.

**Comment by Harrypotterrrr** [comment]
Generally, I think the problem is, just take `avatar.jpg` as an example, jekyll generate `index.html` which contains the assets file with`http://localhost:4000/assets/avatar.jpg:4000` link. But this should be `http://106.13.xx.xx:4000/assets/avatar.jpg:4000`. Because I can successfully get `avatar.jpg` through the second URL, but the first one with `localhost` I failed.

**Comment by iBug** [comment]
You should have used ***local*** forwarding, i.e. `ssh -L`, not `ssh -R`, which would do the reverse. With `ssh -L`, you can visit your local port and SSH will access the remote port.

Check [this question](https://unix.stackexchange.com/q/115897/211239) for better understanding of SSH port forwarding.

**Comment by Harrypotterrrr** [comment]
@iBug Thanks a lot. I used to use remote ssh forwarding to fetch remote resources, eg. code-server, jupyter etc. Besides using remote forward on the inner server with `ssh -NfR`, I run `ssh -NfL 4000:localhost:4000 root@106.13.xx.xx` to forward local to the public remote server. Now everything works well. Appreciate your reply and documentation sincerely.
